{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4e9ada61-3721-4605-8dc3-b2bf0e83fa8a",
   "metadata": {},
   "source": [
    "# Leave-one-out approach"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d189e1e-6f54-4636-aecd-cd976ed66238",
   "metadata": {
    "tags": []
   },
   "source": [
    "Mask SNPs to simulate lower SNP density. The example provided here is for masking 5% of SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c49d2d9-6d83-4781-89b4-a652b64977a3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      target-thin\n",
    "#SBATCH --mem           2G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load modules\n",
    "module load PLINK/2.00a2.3\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "module load vcflib/1.0.1-GCC-9.2.0\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target\n",
    "cd $maindir\n",
    "\n",
    "# subsample target file to SNP density of 2.73K SNPs/Mb\n",
    "vcfrandomsample -r 0.95 \\\n",
    "/nesi/nobackup/project/project_2/imputation-input/ref/hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele.vcf.gz > \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin5.vcf\n",
    "bcftools view hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin5.vcf -Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin5.vcf.gz\n",
    "tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin5.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f57d949-6f9a-4046-87fc-1d9eb82dd5d3",
   "metadata": {},
   "source": [
    "Create target files with one sample selected as the target each time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68eecd97-afc1-417a-ad39-ba169d41fe9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      target-3-extract_sample\n",
    "#SBATCH --mem           20G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load modules\n",
    "module load PLINK/2.00a2.3\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target\n",
    "cd $maindir\n",
    "\n",
    "# extract each sample into its own file\n",
    "while read s; do\n",
    "  bcftools view -s ${s} hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin5.vcf.gz -Oz -o target-${s}-thin5.vcf.gz\n",
    "  tabix -f target-${s}-thin5.vcf.gz\n",
    "done < /nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/ref.list # list of sample names\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78aefe2c-c07f-4113-b3e0-e7b7f33649b6",
   "metadata": {},
   "source": [
    "Create corresponding reference panels for each target sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a848852-4d34-44fa-b483-ace8b6463ac7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      ref-1-make_ref\n",
    "#SBATCH --mem           20G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load modules\n",
    "module load PLINK/2.00a2.3\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/\n",
    "cd $maindir\n",
    "\n",
    "# extract each sample into its own file\n",
    "while read s; do\n",
    "  bcftools view -s ^${s} \\\n",
    "  /nesi/nobackup/project/project_2/imputation-input/ref/hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele.vcf.gz -Oz -o $maindir/ref/ref-${s}.vcf.gz\n",
    "  tabix -f $maindir/ref/ref-${s}.vcf.gz\n",
    "done < /nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/ref.list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a037198c-c3f5-4d3b-974f-37d6e3ba5144",
   "metadata": {},
   "source": [
    "# Imputation tests & evaluation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6b7c64-0317-41d6-8178-fde91230ecd6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Tests of imputation parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e88d695-6afb-431e-b26e-eae2befd0741",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Default parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8a07fa1-c967-44f6-ba66-f1d289a2cd55",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      test-1-default\n",
    "#SBATCH --mem           20G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --array         1-30\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load modules\n",
    "module load Beagle/5.4-22Jul22.46e\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests\n",
    "cd $maindir\n",
    "\n",
    "# reference config file\n",
    "config=/nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/array.config\n",
    "\n",
    "# extract sample name\n",
    "sample=$(awk -v ArrayTaskID=$SLURM_ARRAY_TASK_ID '$1==ArrayTaskID {print $2}' $config)\n",
    "\n",
    "# import files\n",
    "target=$maindir/target/target-${sample}-thin5.vcf.gz\n",
    "refphased=$maindir/ref/ref-${sample}_phased.vcf.gz\n",
    "\n",
    "# impute\n",
    "beagle gt=${target} ref=${refphased} gp=false out=$maindir/test_1-default/target-${sample}-thin5_beagle5-default\n",
    "tabix -f $maindir/test_1-default/target-${sample}-thin5_beagle5-default.vcf.gz\n",
    "\n",
    "# evaluate\n",
    "module purge\n",
    "module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "# load files\n",
    "validate=$maindir/validate/validate-${sample}.vcf.gz\n",
    "imputed=$maindir/test_1-default/target-${sample}-thin5_beagle5-default.vcf.gz\n",
    "\n",
    "# genotype concordance\n",
    "vcf-compare $validate $imputed > $maindir/test_1-default/concordance-${sample}-default\n",
    "\n",
    "# extract dr2\n",
    "bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > $maindir/test_1-default/target-${sample}-thin5_beagle5-default.r2\n",
    "\n",
    "# r2 & IQS\n",
    "module purge\n",
    "module load Python/3.6.3-gimkl-2017a\n",
    "module load cyvcf2\n",
    "cd $maindir/test_1-default\n",
    "python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "--ga $target \\\n",
    "--imputed $imputed \\\n",
    "--wgs $validate \\\n",
    "--ref $refphased\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe7aaf25-dc90-45b5-9626-9be37fdc6441",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Effective population size (<i>Ne<i>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a60a8f-e673-409d-bdfd-25692defb837",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      test_3-ne\n",
    "#SBATCH --mem           20G\n",
    "#SBATCH --time          12:00:00\n",
    "#SBATCH --array         1-10\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "for ne in 10 100 1000 10000 100000 1000000\n",
    "    do\n",
    "    \n",
    "    # load modules\n",
    "    module load Beagle/5.4-22Jul22.46e\n",
    "    module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "    # reference config file\n",
    "    config=/nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/array.config\n",
    "\n",
    "    # extract sample name\n",
    "    sample=$(awk -v ArrayTaskID=$SLURM_ARRAY_TASK_ID '$1==ArrayTaskID {print $2}' $config)\n",
    "\n",
    "    # set working directory\n",
    "    maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/test_3-ne\n",
    "    cd $maindir\n",
    "\n",
    "    # reference panel\n",
    "    ref=/nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/ref-${sample}.vcf.gz\n",
    "    \n",
    "    # phase using Beagle 5\n",
    "    beagle gt=$ref ne=${ne} em=false out=ref-${sample}_phased-ne${ne} nthreads=12\n",
    "    \n",
    "    # index file\n",
    "    tabix -f ref-${sample}_phased-ne${ne}.vcf.gz\n",
    "\n",
    "    # import files\n",
    "    target=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target/target-${sample}-thin5.vcf.gz\n",
    "    refphased=ref-${sample}_phased-ne${ne}.vcf.gz\n",
    "\n",
    "    # impute\n",
    "    beagle gt=${target} ref=${refphased} gp=false ne=${ne} em=false out=target-${sample}-thin5_beagle5-ne${ne}\n",
    "    tabix -f target-${sample}-thin5_beagle5-ne${ne}.vcf.gz\n",
    "\n",
    "    # evaluate\n",
    "    module purge\n",
    "    module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "    module load SAMtools/1.19-GCC-12.3.0\n",
    "    module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "    # load files\n",
    "    validate=/nesi/nobackup/project/project_2/imputation-highCov-allTests/validate/validate-${sample}.vcf.gz\n",
    "    imputed=target-${sample}-thin5_beagle5-ne${ne}.vcf.gz\n",
    "\n",
    "    # genotype concordance\n",
    "    vcf-compare $validate $imputed > concordance-${sample}-ne${ne}\n",
    "\n",
    "    # extract dr2\n",
    "    bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > target-${sample}-thin5_beagle5-ne${ne}.r2\n",
    "\n",
    "    # IQS\n",
    "    module purge\n",
    "    module spider Python/3.9.9-gimkl-2020a\n",
    "    module load cyvcf2\n",
    "    python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "    --ga $target \\\n",
    "    --imputed $imputed \\\n",
    "    --wgs $validate \\\n",
    "    --ref $refphased\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac4f4bc9-ea9c-4ba4-a4fc-2de5a3dcb62c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Combination of parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "298e780c-a453-499d-96a7-5cdc2efa8fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      test_4-combination\n",
    "#SBATCH --mem           40G\n",
    "#SBATCH --time          60:00:00\n",
    "#SBATCH --array         1-30\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "for window in 40 60 80\n",
    "    do\n",
    "    \n",
    "    for iter in 12 24\n",
    "        do\n",
    "        \n",
    "        for impState in 500 1000 1600\n",
    "            do\n",
    "            \n",
    "            for impSeg in 6 20 50\n",
    "                do\n",
    "\n",
    "                # load modules\n",
    "                module load Beagle/5.4-22Jul22.46e\n",
    "                module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "                # reference config file\n",
    "                config=/nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/array.config\n",
    "\n",
    "                # extract sample name\n",
    "                sample=$(awk -v ArrayTaskID=$SLURM_ARRAY_TASK_ID '$1==ArrayTaskID {print $2}' $config)\n",
    "\n",
    "                # set working directory\n",
    "                maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/test_4-combination\n",
    "                cd $maindir\n",
    "\n",
    "                # reference panel\n",
    "                ref=/nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/ref-${sample}.vcf.gz\n",
    "                # phase using Beagle 5\n",
    "                beagle gt=$ref window=${window} iterations=${iter} out=ref-${sample}_phased-w${window}-i${iter} nthreads=12\n",
    "                # index file\n",
    "                tabix -f ref-${sample}_phased-w${window}-i${iter}.vcf.gz\n",
    "\n",
    "                # import files\n",
    "                target=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target/target-${sample}-thin5.vcf.gz\n",
    "                refphased=ref-${sample}_phased-w${window}-i${iter}.vcf.gz\n",
    "\n",
    "                # impute\n",
    "                beagle gt=${target} ref=${refphased} gp=false window=${window} imp-states=${impState} imp-segment=${impSeg} out=target-${sample}-thin5_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}\n",
    "                tabix -f target-${sample}-thin5_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}.vcf.gz\n",
    "\n",
    "                # evaluate\n",
    "                module purge\n",
    "                module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "                module load SAMtools/1.19-GCC-12.3.0\n",
    "                module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "                # load files\n",
    "                validate=/nesi/nobackup/project/project_2/imputation-highCov-allTests/validate/validate-${sample}.vcf.gz\n",
    "                imputed=target-${sample}-thin5_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}.vcf.gz\n",
    "\n",
    "                # genotype concordance\n",
    "                vcf-compare $validate $imputed > concordance-${sample}-w${window}-i${iter}-st${impState}-sg${impSeg}\n",
    "\n",
    "                # extract dr2\n",
    "                bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > target-${sample}-thin5_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}.r2\n",
    "\n",
    "                # IQS\n",
    "                module purge\n",
    "                module spider Python/3.9.9-gimkl-2020a\n",
    "                module load cyvcf2\n",
    "                python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "                --ga $target \\\n",
    "                --imputed $imputed \\\n",
    "                --wgs $validate \\\n",
    "                --ref $refphased\n",
    "                done\n",
    "            done\n",
    "        done\n",
    "    done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "938db0a3-8162-43d3-adfd-43a057839e54",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Inclusion of linkage maps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8bbcfb8-1918-4829-a44e-2bd60efc2157",
   "metadata": {},
   "source": [
    "Example given is for the analysis using a linkage map constructed using CRI-MAP v2.4 and with parameters changed from default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0059b902-3bde-447e-a12f-3b3c397c9785",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      test_2-map-crimap\n",
    "#SBATCH --mem           40G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --array         1-30\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load modules\n",
    "module load Beagle/5.4-22Jul22.46e\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# reference config file\n",
    "config=/nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/array.config\n",
    "\n",
    "# extract sample name\n",
    "sample=$(awk -v ArrayTaskID=$SLURM_ARRAY_TASK_ID '$1==ArrayTaskID {print $2}' $config)\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/test_2-map-crimap\n",
    "cd $maindir\n",
    "\n",
    "# set parameters\n",
    "iterations=24\n",
    "window=60\n",
    "impSeg=20\n",
    "\n",
    "# reference panel\n",
    "ref=/nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/ref-${sample}.vcf.gz\n",
    "\n",
    "# phase using Beagle 5\n",
    "beagle gt=$ref out=ref-${sample}_phased-map-crimap-i${iterations}-w${window} \\\n",
    "nthreads=12 iterations=${iterations} window=${window} \\\n",
    "map=/nesi/nobackup/project/project_2/imputation-input/hihi-autosomes-crimap.map\n",
    "\n",
    "# index file\n",
    "tabix -f ref-${sample}_phased-map-crimap-i${iterations}-w${window}.vcf.gz\n",
    "\n",
    "# import files\n",
    "target=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target/target-${sample}-thin5.vcf.gz\n",
    "refphased=ref-${sample}_phased-map-crimap-i${iterations}-w${window}.vcf.gz\n",
    "\n",
    "# impute\n",
    "beagle gt=${target} ref=${refphased} \\\n",
    "nthreads=12 iterations=${iterations} window=${window} imp-segment=${impSeg} \\\n",
    "map=/nesi/nobackup/project/project_2/imputation-input/hihi-autosomes-crimap.map \\\n",
    "out=target-${sample}-thin5_beagle5-map-crimap-i${iterations}-w${window}-sg${impSeg}\n",
    "tabix -f target-${sample}-thin5_beagle5-map-crimap-i${iterations}-w${window}-sg${impSeg}.vcf.gz\n",
    "\n",
    "# evaluate\n",
    "module purge\n",
    "module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "# load files\n",
    "validate=/nesi/nobackup/project/project_2/imputation-highCov-allTests/validate/validate-${sample}.vcf.gz\n",
    "imputed=target-${sample}-thin5_beagle5-map-crimap-i${iterations}-w${window}-sg${impSeg}.vcf.gz\n",
    "\n",
    "# genotype concordance\n",
    "vcf-compare $validate $imputed > concordance-${sample}-map-crimap-i${iterations}-w${window}-sg${impSeg}\n",
    "\n",
    "# extract dr2\n",
    "bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > target-${sample}-thin5_beagle5-map-crimap-i${iterations}-w${window}-sg${impSeg}.r2\n",
    "\n",
    "# IQS\n",
    "module purge\n",
    "module spider Python/3.9.9-gimkl-2020a\n",
    "module load cyvcf2\n",
    "python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "--ga $target \\\n",
    "--imputed $imputed \\\n",
    "--wgs $validate \\\n",
    "--ref $refphased\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43b74b8d-1cbd-40ce-8a5f-4f1de4aa9650",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Tests of input data structure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff0c011-8761-4075-a5a6-3b443058008f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Exclusion of distant samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc8fa74c-d59d-4872-8b78-2e2d57390154",
   "metadata": {},
   "source": [
    "Calculate genetic distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "290c2999-4053-472e-af56-5918bbadc9f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      vcf-pairwise_relate\n",
    "#SBATCH --mem           10G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load programs\n",
    "module load PLINK/2.00a2.3\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-input\n",
    "cd $maindir\n",
    "\n",
    "# pairwise relatedness\n",
    "plink2 --vcf /nesi/nobackup/project/project_2/imputation-input/ref/hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele.vcf.gz \\\n",
    "--make-king-table --chr-set 95 --allow-extra-chr \\\n",
    "--out hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a4740a0-89c6-49cb-bb84-c2e390ba637f",
   "metadata": {},
   "source": [
    "Prepare reference panels with distant samples excluded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d795699-9e8d-4545-9d5e-e85aee1610b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      ref-distant\n",
    "#SBATCH --mem           20G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load modules\n",
    "module load PLINK/2.00a2.3\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/\n",
    "cd $maindir\n",
    "\n",
    "# remove distant samples (n=5) from ref panel files\n",
    " bcftools view -s ^distant1,distant2,distant3,distant4,distant5 \\\n",
    " /nesi/nobackup/project/project_2/imputation-input/ref/hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele.vcf.gz -Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-distant.vcf.gz\n",
    "  tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-distant.vcf.gz\n",
    "\n",
    "# extract each sample into its own file\n",
    "while read s; do\n",
    "  bcftools view -s ^${s} hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-distant.vcf.gz -Oz -o ref-distant-${s}.vcf.gz\n",
    "  tabix -f ref-distant-${s}.vcf.gz\n",
    "done < /nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/ref-distant.list # list of samples without distant individuals\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7256b195-dc0b-47a4-a100-884d6627ba41",
   "metadata": {
    "tags": []
   },
   "source": [
    "Run imputation test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2c099fc-e00d-4f49-a4ba-cfef8781cfaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      test_5-distant\n",
    "#SBATCH --mem           20G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --array         1-30\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# load modules\n",
    "module load Beagle/5.4-22Jul22.46e\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# reference config file\n",
    "config=/nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/array.config\n",
    "\n",
    "# extract sample name\n",
    "sample=$(awk -v ArrayTaskID=$SLURM_ARRAY_TASK_ID '$1==ArrayTaskID {print $2}' $config)\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/test_5-distant\n",
    "cd $maindir\n",
    "\n",
    "# reference panel\n",
    "ref=/nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/ref-distant-${sample}.vcf.gz\n",
    "\n",
    "# phase using Beagle 5\n",
    "beagle gt=$ref out=ref-distant-${sample}_phased nthreads=12\n",
    "\n",
    "# index file\n",
    "tabix -f ref-distant-${sample}_phased.vcf.gz\n",
    "\n",
    "# import files\n",
    "target=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target/target-${sample}-thin5.vcf.gz\n",
    "refphased=ref-distant-${sample}_phased.vcf.gz\n",
    "\n",
    "# impute\n",
    "beagle gt=${target} ref=${refphased} gp=true out=target-${sample}-thin5_beagle5-distant\n",
    "tabix -f target-${sample}-thin5_beagle5-distant.vcf.gz\n",
    "\n",
    "# evaluate\n",
    "module purge\n",
    "module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "# load files\n",
    "validate=/nesi/nobackup/project/project_2/imputation-highCov-allTests/validate/validate-${sample}.vcf.gz\n",
    "imputed=target-${sample}-thin5_beagle5-distant.vcf.gz\n",
    "\n",
    "# genotype concordance\n",
    "vcf-compare $validate $imputed > concordance-${sample}-distant\n",
    "\n",
    "# extract dr2\n",
    "bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > target-${sample}-thin5_beagle5-distant.r2\n",
    "\n",
    "# IQS\n",
    "module purge\n",
    "module spider Python/3.9.9-gimkl-2020a\n",
    "module load cyvcf2\n",
    "python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "--ga $target \\\n",
    "--imputed $imputed \\\n",
    "--wgs $validate \\\n",
    "--ref $refphased\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef2bb5b0-230b-47a7-a19f-244b62ff2825",
   "metadata": {},
   "source": [
    "### Number of individuals in reference panels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89af6b5a-8f6c-4d26-97b5-91577a454cbd",
   "metadata": {},
   "source": [
    "Individuals were randomly sampled from all 30 high-coverage samples, with 10 repeats per reference panel size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3677804-8a5d-4e4c-8a44-134f4750b787",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      test_8-ref_thin-v2\n",
    "#SBATCH --mem           20G\n",
    "#SBATCH --time          60:00:00\n",
    "#SBATCH --array         1-26\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# reference config file\n",
    "config=/nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/array-Tiritiri.config\n",
    "\n",
    "# extract sample name\n",
    "sample=$(awk -v ArrayTaskID=$SLURM_ARRAY_TASK_ID '$1==ArrayTaskID {print $2}' $config)\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/test_8-ref_thin\n",
    "cd $maindir\n",
    "\n",
    "for j in 5 10 15 20\n",
    "do\n",
    "    for i in {1..10}\n",
    "    do\n",
    "        module purge\n",
    "        module load Beagle/5.4-22Jul22.46e\n",
    "        module load SAMtools/1.19-GCC-12.3.0\n",
    "        module load BCFtools/1.19-GCC-11.3.0\n",
    "        \n",
    "        # number of rows to remove\n",
    "        let remove=29-${j}\n",
    "\n",
    "        # output sample names from reference panel, shuffle order, remove first X rows, re-order (randomly subset X rows)\n",
    "        bcftools query -l /nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/ref-${sample}.vcf.gz | shuf | sed \"1,${remove}d\" | sort > \\\n",
    "        /nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/ref_thin${j}-${sample}_${i}.txt\n",
    "\n",
    "        cd /nesi/nobackup/project/project_2/imputation-highCov-allTests/ref\n",
    "\n",
    "        # subset reference panel and rename with a number from 1-10\n",
    "        bcftools view -S ref_thin${j}-${sample}_${i}.txt \\\n",
    "        ref-${sample}.vcf.gz -Oz -o ref_thin${j}-${sample}_${i}.vcf.gz\n",
    "        tabix -f ref_thin${j}-${sample}_${i}.vcf.gz\n",
    "\n",
    "        cd $maindir\n",
    "\n",
    "        # reference panel\n",
    "        ref=/nesi/nobackup/project/project_2/imputation-highCov-allTests/ref/ref_thin${j}-${sample}_${i}.vcf.gz\n",
    "        # phase using Beagle 5\n",
    "        beagle gt=$ref out=ref_thin${j}-${sample}_${i}_phased nthreads=12\n",
    "        # index file\n",
    "        tabix -f ref_thin${j}-${sample}_${i}_phased.vcf.gz\n",
    "\n",
    "        # import files\n",
    "        target=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target/target-${sample}-thin5.vcf.gz\n",
    "        refphased=ref_thin${j}-${sample}_${i}_phased.vcf.gz\n",
    "\n",
    "        # impute\n",
    "        beagle gt=${target} ref=${refphased} gp=true out=target-${sample}-thin5_beagle5-ref_thin${j}_${i}\n",
    "        tabix -f target-${sample}-thin5_beagle5-ref_thin${j}_${i}.vcf.gz\n",
    "\n",
    "        # evaluate\n",
    "        module purge\n",
    "        module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "        module load SAMtools/1.19-GCC-12.3.0\n",
    "        module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "        # load files\n",
    "        validate=/nesi/nobackup/project/project_2/imputation-highCov-allTests/validate/validate-${sample}.vcf.gz\n",
    "        imputed=target-${sample}-thin5_beagle5-ref_thin${j}_${i}.vcf.gz\n",
    "\n",
    "        # genotype concordance\n",
    "        vcf-compare $validate $imputed > concordance-${sample}-ref_thin${j}_${i}\n",
    "\n",
    "        # extract dr2\n",
    "        bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > target-${sample}-thin5_beagle5-ref_thin${j}_${i}.r2\n",
    "\n",
    "        # IQS\n",
    "        module purge\n",
    "        module spider Python/3.9.9-gimkl-2020a\n",
    "        module load cyvcf2\n",
    "        python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "        --ga $target \\\n",
    "        --imputed $imputed \\\n",
    "        --wgs $validate \\\n",
    "        --ref $refphased\n",
    "\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8403c8a-7ffd-43ff-8191-215f35320558",
   "metadata": {},
   "source": [
    "### SNP densities"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7045dee-09eb-45e2-a949-7baecc7da7d3",
   "metadata": {},
   "source": [
    "Subsample SNPs in target file to simulate lower SNP densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85d061d8-3b9e-4c16-b9f5-d36d089d2c49",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      target-thin\n",
    "#SBATCH --mem           2G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH --account       project\n",
    "#SBATCH --partition=milan\n",
    "#SBATCH --output=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target/logs/%x-%j.out\n",
    "\n",
    "# load modules\n",
    "module load PLINK/2.00a2.3\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "module load vcflib/1.0.1-GCC-9.2.0\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target\n",
    "cd $maindir\n",
    "\n",
    "# subsample target file to SNP density of 2.10K SNPs/Mb\n",
    "vcfrandomsample -r 0.77 \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin5.vcf.gz > \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin27.vcf\n",
    "bcftools view hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin27.vcf \\\n",
    "-Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin27.vcf.gz\n",
    "tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin27.vcf.gz\n",
    "\n",
    "# subsample target file to SNP density of 1.58K SNPs/Mb\n",
    "vcfrandomsample -r 0.75 \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin27.vcf.gz > \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin45.vcf\n",
    "bcftools view hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin45.vcf \\\n",
    "-Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin45.vcf.gz\n",
    "tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin45.vcf.gz\n",
    "\n",
    "# subsample target file to SNP density of 1.09K SNPs/Mb\n",
    "vcfrandomsample -r 0.69 \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin45.vcf.gz > \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin62.vcf\n",
    "bcftools view hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin62.vcf \\\n",
    "-Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin62.vcf.gz\n",
    "tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin62.vcf.gz\n",
    "\n",
    "# subsample target file to SNP density of 197 SNPs/Mb\n",
    "vcfrandomsample -r 0.18 \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin62.vcf.gz > \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin93.vcf\n",
    "bcftools view hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin93.vcf \\\n",
    "-Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin93.vcf.gz\n",
    "tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin93.vcf.gz\n",
    "\n",
    "# subsample target file to SNP density of 48.9 SNPs/Mb (SNP-chip)\n",
    "vcfrandomsample -r 0.25 \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin5.vcf.gz > \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinSnpchip.vcf\n",
    "bcftools view hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinSnpchip.vcf \\\n",
    "-Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinSnpchip.vcf.gz\n",
    "tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinSnpchip.vcf.gz\n",
    "\n",
    "# subsample target file to SNP density of 19.6 SNPs/Mb (RADseq)\n",
    "vcfrandomsample -r 0.4 \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinSnpchip.vcf.gz > \\\n",
    "hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinRadseq.vcf\n",
    "bcftools view hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinRadseq.vcf \\\n",
    "-Oz -o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinRadseq.vcf.gz\n",
    "tabix -f hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thinRadseq.vcf.gz\n",
    "\n",
    "# extract each sample into its own file\n",
    "while read s; do\n",
    "    for sub in 27 45 62 93 Snpchip Radseq\n",
    "    do\n",
    "        bcftools view -s ${s} hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-thin${sub}.vcf.gz -Oz -o target-${s}-thin${sub}.vcf.gz\n",
    "        tabix -f target-${s}-thin${sub}.vcf.gz\n",
    "        bcftools stats -s \"-\" target-${s}-thin${sub}.vcf.gz > target-${s}-thin${sub}.stats\n",
    "        done\n",
    "done < /nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/ref.list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "464be446-49a9-44b2-92b8-7d7daa319984",
   "metadata": {},
   "source": [
    "Run imputation test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc6c3cce-9695-4e29-a787-11753fa27b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      test_7-thin-gp\n",
    "#SBATCH --mem           40G\n",
    "#SBATCH --time          15:00:00\n",
    "#SBATCH --array         1-30\n",
    "#SBATCH --cpus-per-task 12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# reference config file\n",
    "config=/nesi/nobackup/project/project_2/imputation-highCov-allTests/codes/array.config\n",
    "\n",
    "# extract sample name\n",
    "sample=$(awk -v ArrayTaskID=$SLURM_ARRAY_TASK_ID '$1==ArrayTaskID {print $2}' $config)\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-highCov-allTests/test_7-thin-gp\n",
    "cd $maindir\n",
    "\n",
    "for sub in 93 Snpchip Radseq\n",
    "do\n",
    "    for window in 60\n",
    "    do\n",
    "\n",
    "        for iter in 24\n",
    "        do\n",
    "\n",
    "            for impState in 1600\n",
    "            do\n",
    "\n",
    "                for impSeg in 6\n",
    "                do\n",
    "                    module load Beagle/5.4-22Jul22.46e\n",
    "                    module load SAMtools/1.19-GCC-12.3.0\n",
    "                    # import files\n",
    "                    target=/nesi/nobackup/project/project_2/imputation-highCov-allTests/target/target-${sample}-thin${sub}.vcf.gz\n",
    "                    refphased=/nesi/nobackup/project/project_2/imputation-highCov-allTests/test_4-combination/ref-${sample}_phased-w${window}-i${iter}.vcf.gz\n",
    "\n",
    "                    # impute\n",
    "                    beagle gt=${target} ref=${refphased} gp=true window=${window} imp-states=${impState} imp-segment=${impSeg} out=target-${sample}-thin${sub}_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}-gp\n",
    "                    tabix -f target-${sample}-thin${sub}_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}-gp.vcf.gz\n",
    "\n",
    "                    # evaluate\n",
    "                    module purge\n",
    "                    module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "                    module load SAMtools/1.19-GCC-12.3.0\n",
    "                    module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "                    # load files\n",
    "                    validate=/nesi/nobackup/project/project_2/imputation-highCov-allTests/validate/validate-${sample}.vcf.gz\n",
    "                    imputed=target-${sample}-thin${sub}_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}-gp.vcf.gz\n",
    "\n",
    "                    # genotype concordance\n",
    "                    vcf-compare $validate $imputed > concordance-${sample}-thin${sub}-w${window}-i${iter}-st${impState}-sg${impSeg}-gp\n",
    "\n",
    "                    # extract dr2\n",
    "                    bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > target-${sample}-thin${sub}_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}-gp.r2\n",
    "\n",
    "                    # IQS\n",
    "                    module purge\n",
    "                    module spider Python/3.9.9-gimkl-2020a\n",
    "                    module load cyvcf2\n",
    "                    python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "                    --ga $target \\\n",
    "                    --imputed $imputed \\\n",
    "                    --wgs $validate \\\n",
    "                    --ref $refphased\n",
    "                done\n",
    "            done\n",
    "        done\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abb2b6ef-653d-4c5e-9579-49d9ceba979d",
   "metadata": {},
   "source": [
    "### Sporadic missingness"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c150c34a-33a5-4fba-af36-08d7c3330efe",
   "metadata": {},
   "source": [
    "Simulate sporadic missingness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50be62d3-2667-494f-9eac-353cd72d6c6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set working directory\n",
    "cd /nesi/nobackup/project/project_2/imputation-wgs-ref-sporadic-highCov\n",
    "\n",
    "# rename samples in target\n",
    "bcftools reheader --samples ./codes/rename-samples.txt \\\n",
    "/nesi/nobackup/project/project_2/imputation-input/ref/hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele.vcf.gz \\\n",
    "-o hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-renameSamples.vcf\n",
    "\n",
    "# compress file\n",
    "bgzip hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-renameSamples.vcf\n",
    "\n",
    "# downsample file to 69K/788K\n",
    "# 69K\n",
    "bcftools view --include ID==@./codes/target-69k.txt hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-renameSamples.vcf.gz -Oz -o target-69K.vcf.gz\n",
    "plink2 --vcf target-69K.vcf.gz --missing --chr-set 95 --allow-extra-chr --out target-69K\n",
    "\n",
    "# 788K\n",
    "bcftools view --include ID==@./codes/target-788k.txt hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-renameSamples.vcf.gz -Oz -o target-788K.vcf.gz\n",
    "plink2 --vcf target-788K.vcf.gz --missing --chr-set 95 --allow-extra-chr --out target-788K\n",
    "\n",
    "## input sporadic missingness using R code (example shown for 788K SNPs allowing 50% missingness)\n",
    "library(pgirmess)\n",
    "library(dplyr)\n",
    "\n",
    "# set working directory\n",
    "setwd(\"/nesi/nobackup/project/project_2/imputation-wgs-ref-sporadic-highCov\")\n",
    "\n",
    "# read in vcf file\n",
    "vcf <- read.table(\"target-788K.vcf\", header = FALSE, fill = TRUE)\n",
    "\n",
    "# trim down to just genotypes\n",
    "vcf2 <- vcf[,-c(1:9)]\n",
    "\n",
    "# matrix size\n",
    "num_rows <- nrow(vcf2)\n",
    "num_cols <- ncol(vcf2)\n",
    "\n",
    "# calculate the total number of cells\n",
    "total_cells <- num_rows * num_cols\n",
    "\n",
    "# set missingness threshold\n",
    "missingThreshold <- 0.5\n",
    "missingPc <-missingThreshold*100\n",
    "\n",
    "# Calculate the number of cells to set as mising (around X%/0.X)\n",
    "num_cells_to_replace <- round(missingThreshold * total_cells)\n",
    "\n",
    "# Create a matrix of random numbers\n",
    "random_matrix <- matrix(sample(c(1:10), total_cells, replace = TRUE), nrow = num_rows, ncol = num_cols)\n",
    "\n",
    "# Flatten the matrix into a vector\n",
    "random_vector <- as.vector(random_matrix)\n",
    "\n",
    "# Randomly select cells set as missing\n",
    "cells_to_replace <- sample(total_cells, num_cells_to_replace)\n",
    "\n",
    "# Set selected cells as missing\n",
    "random_vector[cells_to_replace] <- \"./.\"\n",
    "\n",
    "# Reshape the vector back into a matrix\n",
    "final_matrix <- matrix(random_vector, nrow = num_rows, ncol = num_cols)\n",
    "\n",
    "# Convert matrix to dataframe\n",
    "final_df <- as.data.frame(final_matrix)\n",
    "\n",
    "# If cell value is not equal to \"./.\", grab value from original genotype vcf\n",
    "final_df[final_df != \"./.\"] <- vcf2[final_df != \"./.\"]\n",
    "\n",
    "# Put first few columns back in\n",
    "final_vcf <- cbind(vcf[,c(1:9)], final_df)\n",
    "\n",
    "# output vcf\n",
    "write.delim(final_vcf, file = \"target-788K-sporadic50.vcf\", row.names = FALSE, quote = FALSE, sep = \"\\t\")\n",
    "\n",
    "# compress sporadic file\n",
    "bgzip -c target-69K-sporadic40.vcf > target-69K-sporadic40.vcf.gz\n",
    "tabix -f target-69K-sporadic40.vcf.gz\n",
    "bgzip -c target-788K-sporadic50.vcf > target-788K-sporadic50.vcf.gz\n",
    "tabix -f target-788K-sporadic50.vcf.gz\n",
    "\n",
    "# calculate new missingness\n",
    "plink2 --vcf target-69K-sporadic40.vcf.gz --missing --chr-set 95 --allow-extra-chr --out target-69K-sporadic40\n",
    "plink2 --vcf target-788K-sporadic50.vcf.gz --missing --chr-set 95 --allow-extra-chr --out target-788K-sporadic50\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da87a4ae-d5a2-40b5-af35-8f21820ac192",
   "metadata": {},
   "source": [
    "Run imputation test. Example shown for 788K SNPs allowing 50% missingness."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34f15cb6-b9e1-426e-910e-3ecc66c3c267",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --job-name      impute-ne-788K-sporadic\n",
    "#SBATCH --mem           40G\n",
    "#SBATCH --time          5:00:00\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH --account       project\n",
    "\n",
    "# set up environment\n",
    "module load Beagle/5.4-22Jul22.46e\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# set parameters\n",
    "window=60\n",
    "iter=24\n",
    "impState=1600\n",
    "impSeg=6\n",
    "\n",
    "# set working directory\n",
    "maindir=/nesi/nobackup/project/project_2/imputation-wgs-ref-sporadic-highCov\n",
    "cd $maindir\n",
    "\n",
    "# import files\n",
    "target=target-788K-sporadic50-remove25A.vcf.gz\n",
    "refphased=ref-noMissing_phased-w${window}-i${iter}-remove25A.vcf.gz\n",
    "\n",
    "# impute\n",
    "beagle gt=$target ref=$refphased \\\n",
    "impute=true gp=false window=${window} imp-states=${impState} imp-segment=${impSeg} nthreads=12 \\\n",
    "out=target-788K-sporadic50-remove25A_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}\n",
    "tabix -f target-788K-sporadic50-remove25A_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}.vcf.gz\n",
    "\n",
    "# EVALUATE\n",
    "module purge\n",
    "module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "module load BCFtools/1.19-GCC-11.3.0\n",
    "\n",
    "# load files\n",
    "refphased=ref-noMissing_phased-w${window}-i${iter}-remove25A.vcf.gz\n",
    "target=target-788K-sporadic50-remove25A.vcf.gz\n",
    "validate=hihi_wgs_filter_highcov-autosomes-snpID-noMissing-noSingleton-2allele-renameSamples.vcf.gz\n",
    "imputed=target-788K-sporadic50-remove25A_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}.vcf.gz\n",
    "\n",
    "# genotype concordance\n",
    "vcf-compare $validate $imputed > target-788K-sporadic50-remove25A_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}\n",
    "\n",
    "# extract dr2\n",
    "bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%QUAL\\t%FILTER\\t%DR2\\t%AF\\t%IMP\\n' $imputed > target-788K-sporadic50-remove25A_beagle5-w${window}-i${iter}-st${impState}-sg${impSeg}.r2\n",
    "\n",
    "# IQS\n",
    "module purge\n",
    "module load Python/3.6.3-gimkl-2017a\n",
    "module load cyvcf2\n",
    "python3 /home/user/imputation/Compare_imputation_to_WGS.py \\\n",
    "--ga $target \\\n",
    "--imputed $imputed \\\n",
    "--wgs $validate \\\n",
    "--ref $refphased\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "",
   "name": ""
  },
  "language_info": {
   "name": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
